### ISSUE TITLE
Debouncing and Throttling Explained Through Examples 

### ISSUE THREAD

**Issue by nirjan-dev** [nirjan-dev/learning-to-code#364]
[link to the post](https://css-tricks.com/debouncing-throttling-explained-examples/)

Having a debounced or throttled version of our function is especially useful when we are attaching the function to a DOM event. Why? Because we are giving ourselves a layer of control between the event and the execution of the function. Remember, we don’t control how often those DOM events are going to be emitted. It can vary.

When scrolling using a trackpad, scroll wheel, or just by dragging a scrollbar can trigger easily 30 events per second. But scrolling slowly (swapping) in a smartphone could trigger as much as 100 events per second during my tests. Is your scroll handler prepared for this rate of execution?

# debounce
The Debounce technique allow us to “group” multiple sequential calls in a single one. 

![image](https://user-images.githubusercontent.com/29270715/81398097-49cbfe80-9148-11ea-8f21-e5558bdfc606.png)

You can see how sequential fast events are represented by a single debounced event. But if the events are triggered with big gaps, the debouncing doesn’t happen. 

You may find it irritating that the debouncing event waits before triggering the function execution, until the events stop happening so rapidly. Why not trigger the function execution immediately, so it behaves exactly as the original non-debounced handler? But not fire again until there is a pause in the rapid calls.

You can do this! Here’s an example with the leading flag on:

![image](https://user-images.githubusercontent.com/29270715/81398146-636d4600-9148-11ea-9ff6-985aeffbb016.png)

# How to use debounce and throttle and common pitfalls
 If you only need the _.debounce and _.throttle functions, you can use Lodash custom builder to output a custom 2KB minified library. Build it with this simple command:

```
npm i -g lodash-cli
lodash include = debounce, throttle
```

That said, most use the modular form `lodash/throttle` and `lodash/debounce` or `lodash.throttle` and `lodash.debounce` packages with webpack/browserify/rollup.

A common pitfall is to call the _.debounce function more than once:

```js
// WRONG
$(window).on('scroll', function() {
   _.debounce(doSomething, 300); 
});

// RIGHT
$(window).on('scroll', _.debounce(doSomething, 200));
```

Creating a variable for the debounced function will allow us to call the private method debounced_version.cancel(), available in lodash and underscore.js, in case you need it.

```js
var debounced_version = _.debounce(doSomething, 200);
$(window).on('scroll', debounced_version);

// If you need it
debounced_version.cancel();
```

# throttle

By using _.throttle, we don’t allow to our function to execute more than once every X milliseconds.

The main difference between this and debouncing is that throttle guarantees the execution of the function regularly, at least every X milliseconds.

# requestAnimationFrame (rAF)

requestAnimationFrame is another way of rate-limiting the execution of a function.

It can be thought as a _.throttle(dosomething, 16). But with a much higher fidelity, since it’s a browser native API that aims for better accuracy.

## Pros

   - Aims for 60fps (frames of 16 ms) but internally will decide the best timing on how to schedule the rendering.
   - Fairly simple and standard API, not changing in the future. Less maintenance.

# Cons

  -  The start/cancelation of rAFs it’s our responsibility, unlike .debounce or .throttle, where it’s managed internally.
  -  If the browser tab is not active, it would not execute. Although for scroll, mouse or keyboard events this doesn’t matter.
  -  Although all modern browsers offer rAF, still is not supported in IE9, Opera Mini and old Android. A polyfill would be needed still today.
  -  rAF is not supported in node.js, so you can’t use it on the server to throttle filesystem events.

As a rule of thumb, I would use requestAnimationFrame if your JavaScript function is “painting” or animating directly properties, use it at everything that involves re-calculating element positions.

To make Ajax requests, or deciding if adding/removing a class (that could trigger a CSS animation), I would consider _.debounce or _.throttle, where you can set up lower executing rates (200ms for example, instead of 16ms)

In summary:

  -  debounce: Grouping a sudden burst of events (like keystrokes) into a single one.
  -  throttle: Guaranteeing a constant flow of executions every X milliseconds. Like checking every 200ms your scroll position to trigger a CSS animation.
  -  requestAnimationFrame: a throttle alternative. When your function recalculates and renders elements on screen and you want to guarantee smooth changes or animations. Note: no IE9 support.
